Refactor ai utils

Anna-Xiong opened the pull request:
## Description

## Screenshot

## Notice

- [ ] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

huong-li-nguyen commented on vizro-ai/src/vizro_ai/_vizro_ai.py:
```diff
@@ -123,7 +118,7 @@ def plot(
 
         if code_string.startswith("Failed to debug code"):
-            raise DebugFailure(
+            raise Exception(
```
Don't you want to continue using `DebugFailure` and import it from the`helper.py`? Is it used elsewhere or just here?

huong-li-nguyen commented on vizro-ai/src/vizro_ai/_vizro_ai.py:
```diff
@@ -136,75 +131,3 @@ def plot(
-    code_string: str, max_debug_retry: int, fix_chain: Callable, df: pd.DataFrame = None
-) -> Dict[bool, str]:
-    """Debugging helper."""
-    # TODO plug logic back into component
```
I'll leave it up to you whether you want to change this, but if I would join the project I would have no idea what this TODO is about 😄 I think adding more comments on TODOs or explanatory comments are always nice even if it's lengthy.

Anna-Xiong replied:
This is not applicable anymore, let me clean up the todos here

huong-li-nguyen commented on vizro-ai/src/vizro_ai/_vizro_ai.py:
```diff
@@ -136,75 +131,3 @@ def plot(
-    for attempt in range(max_debug_retry):
-        try:
-            _exec_code(code=code_string, local_args={"df": df}, is_notebook_env=_is_jupyter())
-            retry_success = True
```
Do we want to move the `_is_jupyter()` check outside the loop? I assume it will not change and only needs to be called once? Otherwise leave in

huong-li-nguyen commented on vizro-ai/src/vizro_ai/_vizro_ai.py:
```diff
@@ -136,75 +131,3 @@ def plot(
-
-    if show_fig and "\nfig.show()" not in code:
-        code += "\nfig.show()"
-    elif not show_fig:
```
Can this be converted to a simple else? If not is there an else missing? It can also be a simple `else: pass` just for readibility?

huong-li-nguyen commented on vizro-ai/src/vizro_ai/_vizro_ai.py:
```diff
@@ -136,75 +131,3 @@ def plot(
-
-def _display_markdown_and_chart(df: pd.DataFrame, code_snippet: str, biz_insights: str, code_explain: str) -> None:
-    # TODO change default test str to other
-    """Display chart and Markdown format description in jupyter."""
-    try:
-        # pylint: disable=import-outside-toplevel
-        from IPython.display import Markdown, display
-    except Exception as exc:
-        raise ImportError("Please install IPython before proceeding in jupyter environment.") from exc
```
Shall we put this in its own function `_check_ipython_dependency` or similar?

Anna-Xiong replied:
that might make sense, only thing is that if we move this out, would we need to handle the import error here separately in this function for `display` and `Markdown`, seems duplicate work

huong-li-nguyen replied:
Leave it as is then :)

huong-li-nguyen commented on vizro-ai/src/vizro_ai/_vizro_ai.py:
````diff
@@ -136,75 +131,3 @@ def plot(
-        raise ImportError("Please install IPython before proceeding in jupyter environment.") from exc
-    # TODO clean up the formatting markdown code
-    markdown_code = f"```\n{code_snippet}\n```"
-    output_text = f"<h4>Insights:</h4>\n\n{biz_insights}\n<br><br><h4>Code:</h4>\n\n{code_explain}\n{markdown_code}"
-    display(Markdown(output_text))
````
Can't this markdown string be formatted as below? Similar to how we have our card? Or is that not possible:

```
f"""
#### Insights:

{biz_insights}

#### Code:

{code_explain}

"""
```

Anna-Xiong replied:
unfortunately back then the normal markdown format does not work for jupyter to render it correctly, and we came up with those combination to finally make it work. Let's leave it for now but keep the TODOs here

huong-li-nguyen reviewed (approved):
Hey @Anna-Xiong  - thanks for tidying up - I always appreciate that!

Have some minor comments as I've just been reading through this code for the first time. Really up to you, whether you want to do a bit more refactoring while you're already on it, or if you just want to keep the PR as moving code to other files. I am fine either way 👍 

It would help if the comments/todos have a bit more context? Just in case someone else would pick this up 😄

Anna-Xiong commented:
> Hey @Anna-Xiong - thanks for tidying up - I always appreciate that!
> 
> Have some minor comments as I've just been reading through this code for the first time. Really up to you, whether you want to do a bit more refactoring while you're already on it, or if you just want to keep the PR as moving code to other files. I am fine either way 👍
> 
> It would help if the comments/todos have a bit more context? Just in case someone else would pick this up 😄

thank you Li! Most of this pr is to move helpers so that new pr can reuse those functions, but definitely worth addressing those comments while we can. i have addressed some of them and probably two of them need a bit more discussion/time to refine.
